File: packages/sdk-ts/src/robotics/provenance.ts
/**
 * Model-and-config provenance attestation for robots (Phase 5.2), TypeScript.
 *
 * Mirrors `vouch/robotics/provenance.py` with byte-identical output. A
 * ModelProvenanceAttestation is an eddsa-jcs-2022 VC recording the VLA model
 * name, weights hash, safety policy, and a configHash computed over the
 * JCS-canonical config so any verifier reproduces it. Re-signable on an OTA
 * update via `supersedes`, forming a tamper-evident chain.
 */

import * as crypto from 'crypto';

import { verifyProof } from '../data-integrity';
import { canonicalize } from '../jcs';
import type { Signer } from '../signer';
import { VC_CONTEXT_V2, VOUCH_CONTEXT_V1 } from '../vc';

export const MODEL_PROVENANCE_TYPE = 'ModelProvenanceAttestation';

function iso(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}

/** Multibase SHA-256 of the JCS-canonical config object. */
export function configHash(config: Record<string, unknown>): string {
  const digest = crypto.createHash('sha256').update(canonicalize(config)).digest();
  return 'u' + digest.toString('base64url');
}

export interface BuildProvenanceOptions {
  robotDid: string;
  modelName: string;
  weightsHash: string;
  safetyPolicy: string;
  config?: Record<string, unknown>;
  version?: string;
  supersedes?: string;
  validSeconds?: number;
  validFrom?: Date;
}

/** Build a signed ModelProvenanceAttestation for the software on a robot. */
export async function buildProvenanceAttestation(
  signer: Signer,
  opts: BuildProvenanceOptions
): Promise<Record<string, unknown>> {
  const issued = opts.validFrom ?? new Date();
  const vla: Record<string, unknown> = {
    modelName: opts.modelName,
    weightsHash: opts.weightsHash,
    safetyPolicy: opts.safetyPolicy,
  };
  if (opts.version !== undefined) vla.version = opts.version;
  if (opts.config !== undefined) vla.configHash = configHash(opts.config);

  const subject: Record<string, unknown> = { id: opts.robotDid, vla };
  if (opts.supersedes !== undefined) subject.supersedes = opts.supersedes;

  const credential: Record<string, unknown> = {
    '@context': [VC_CONTEXT_V2, VOUCH_CONTEXT_V1],
    type: ['VerifiableCredential', MODEL_PROVENANCE_TYPE],
    issuer: signer.getDid(),
    validFrom: iso(issued),
    credentialSubject: subject,
  };
  if (opts.validSeconds !== undefined) {
    credential.validUntil = iso(new Date(issued.getTime() + opts.validSeconds * 1000));
  }
  return signer.attachProof(credential);
}

/**
 * Verify a ModelProvenanceAttestation. When `config` is supplied, also check
 * that its hash matches the recorded configHash.
 */
export function verifyProvenanceAttestation(
  attestation: Record<string, unknown>,
  publicKey: crypto.KeyObject,
  config?: Record<string, unknown>
): { ok: boolean; subject?: Record<string, unknown> } {
  const typeField = attestation.type;
  const types = Array.isArray(typeField) ? typeField : [typeField];
  if (!types.includes(MODEL_PROVENANCE_TYPE)) return { ok: false };

  try {
    if (!verifyProof(attestation, publicKey)) return { ok: false };
  } catch {
    return { ok: false };
  }

  const subject = (attestation.credentialSubject ?? {}) as Record<string, unknown>;
  if (config !== undefined) {
    const vla = (subject.vla ?? {}) as Record<string, unknown>;
    if (vla.configHash !== configHash(config)) return { ok: false };
  }
  return { ok: true, subject };
}

File: go-sidecar/robotics/provenance.go
// Model-and-config provenance attestation for robots (Phase 5.2), Go.
//
// Mirrors vouch/robotics/provenance.py and the TypeScript SDK with byte-identical
// output. A ModelProvenanceAttestation is an eddsa-jcs-2022 VC recording the VLA
// model name, weights hash, safety policy, and a configHash computed over the
// JCS-canonical config so any verifier reproduces it. It is re-signable on an OTA
// update via supersedes, forming a tamper-evident chain.
package robotics

import (
	"crypto/ed25519"
	"crypto/sha256"
	"time"

	"github.com/vouch-protocol/vouch/go-sidecar/signer"
)

// ModelProvenanceType is the credential type for a provenance attestation.
const ModelProvenanceType = "ModelProvenanceAttestation"

// ConfigHash returns the multibase SHA-256 of the JCS-canonical config object.
// Python, TypeScript, and Go all canonicalize identically, so the digest is the
// same byte string in every language.
func ConfigHash(config map[string]any) (string, error) {
	canon, err := signer.Canonicalize(config)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256(canon)
	return mb64(sum[:]), nil
}

// BuildProvenanceOptions configures BuildProvenanceAttestation. RobotDID names
// the robot the attestation is about; the issuer is the signer (the robot itself
// or an authority attesting on its behalf).
type BuildProvenanceOptions struct {
	RobotDID     string
	ModelName    string
	WeightsHash  string
	SafetyPolicy string
	Config       map[string]any // optional; nil omits configHash
	Version      string         // optional; "" omits the field
	Supersedes   string         // optional; "" omits the field (prior attestation id)
	ValidSeconds int            // 0 omits validUntil
	ValidFrom    time.Time      // zero uses now
}

// BuildProvenanceAttestation builds a signed ModelProvenanceAttestation for the
// software running on a robot.
func BuildProvenanceAttestation(s *signer.Signer, opts BuildProvenanceOptions) (map[string]any, error) {
	issued := opts.ValidFrom
	if issued.IsZero() {
		issued = time.Now().UTC()
	}

	vla := map[string]any{
		"modelName":    opts.ModelName,
		"weightsHash":  opts.WeightsHash,
		"safetyPolicy": opts.SafetyPolicy,
	}
	if opts.Version != "" {
		vla["version"] = opts.Version
	}
	if opts.Config != nil {
		ch, err := ConfigHash(opts.Config)
		if err != nil {
			return nil, err
		}
		vla["configHash"] = ch
	}

	subject := map[string]any{"id": opts.RobotDID, "vla": vla}
	if opts.Supersedes != "" {
		subject["supersedes"] = opts.Supersedes
	}

	cred := map[string]any{
		"@context":          []any{vcContextV2, vouchContextV1},
		"type":              []any{"VerifiableCredential", ModelProvenanceType},
		"issuer":            s.DID(),
		"validFrom":         iso(issued),
		"credentialSubject": subject,
	}
	if opts.ValidSeconds > 0 {
		cred["validUntil"] = iso(issued.Add(time.Duration(opts.ValidSeconds) * time.Second))
	}
	return s.AttachProof(cred)
}

// VerifyProvenanceAttestation verifies a ModelProvenanceAttestation proof. When
// config is non-nil, it also checks that the recorded configHash reproduces from
// that config. Returns (ok, credentialSubject).
func VerifyProvenanceAttestation(att map[string]any, pub ed25519.PublicKey, config map[string]any) (bool, map[string]any) {
	if !hasType(att["type"], ModelProvenanceType) {
		return false, nil
	}
	ok, err := signer.VerifyDataIntegrityProof(att, pub)
	if err != nil || !ok {
		return false, nil
	}

	subject, _ := att["credentialSubject"].(map[string]any)
	if config != nil {
		vla, _ := subject["vla"].(map[string]any)
		want, err := ConfigHash(config)
		if err != nil {
			return false, nil
		}
		if got, _ := vla["configHash"].(string); got != want {
			return false, nil
		}
	}
	return true, subject
}
